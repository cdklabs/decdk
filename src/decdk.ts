import * as cdk from 'aws-cdk-lib';
import chalk from 'chalk';
import yargs from 'yargs';
import { DeclarativeStack, loadTypeSystem, readTemplate, stackNameFromFileName } from '.';

async function main() {
  const argv = await yargs
    .usage('$0 <filename>', 'Synthesize a CDK stack from a declarative JSON schema')
    .positional('filename', { type: 'string', required: true })
    .argv;


  const templateFile = argv.filename;
  if (!templateFile) {
    throw new Error('filename is missing');
  }

  const template = await readTemplate(templateFile);
  const stackName = stackNameFromFileName(templateFile);
  const typeSystem = await loadTypeSystem();

  const app = new cdk.App();
  new DeclarativeStack(app, stackName, { template, typeSystem });
  app.synth();
}

main().catch(e => {
  // eslint-disable-next-line no-console
  console.error(chalk.red(e));
  process.exit(1);
});
