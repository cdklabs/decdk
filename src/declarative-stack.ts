import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as reflect from 'jsii-reflect';
import { AnnotationsContext, DeclarativeStackError } from './error-handling';
import { EvaluationContext, Evaluator } from './evaluate';
import { Template } from './parser/template';
import { TypedTemplate } from './type-resolution/template';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const { version } = require('../package.json');

const JSII_RTTI_SYMBOL = Symbol.for('jsii.rtti');

export interface DeclarativeStackProps extends cdk.StackProps {
  typeSystem: reflect.TypeSystem;
  template: Template;
  workingDirectory?: string;
}

export class DeclarativeStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: DeclarativeStackProps) {
    super(scope, id, props);

    const ctx = props.template.metadata.get('AWS::CDK::Context') ?? {};
    Object.entries(ctx).forEach(([k, v]) => this.node.setContext(k, v));

    this.templateOptions.templateFormatVersion =
      props.template.templateFormatVersion;
    this.templateOptions.description = props.template.description;

    const typeSystem = props.typeSystem;
    const template = new TypedTemplate(props.template, { typeSystem });

    new (class extends Construct {
      //@ts-ignore
      private static readonly [JSII_RTTI_SYMBOL] = {
        fqn: '@cdklabs/decdk',
        version,
      };
    })(this, '$decdkAnalytics');

    const annotations = AnnotationsContext.root();
    const context = new EvaluationContext({
      stack: this,
      template,
      typeSystem,
    });
    const ev = new Evaluator(context);

    _cwd(props.workingDirectory, () => {
      ev.evaluateTemplate(annotations);
    });

    if (annotations.hasErrors()) {
      throw new DeclarativeStackError(annotations);
    }
  }
}

function _cwd<T>(workDir: string | undefined, cb: () => T): T {
  if (!workDir) {
    return cb();
  }
  const prevWd = process.cwd();
  try {
    process.chdir(workDir);
    return cb();
  } finally {
    process.chdir(prevWd);
  }
}
