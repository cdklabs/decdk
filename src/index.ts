export * from './declarative-stack';
export * from './util';
