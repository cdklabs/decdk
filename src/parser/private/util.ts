export function mkDict<A>(
  xs: ReadonlyArray<readonly [string, A]>
): Record<string, A> {
  const ret: Record<string, A> = {};
  for (const [k, v] of xs) {
    ret[k] = v;
  }
  return ret;
}
