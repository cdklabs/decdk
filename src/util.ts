import * as path from 'path';
import * as jsiiReflect from 'jsii-reflect';
import { Template } from './parser/template';

/**
 * Reads a YAML/JSON template file.
 */
export async function readTemplate(templateFile: string): Promise<Template> {
  return Template.fromFile(templateFile);
}

export async function loadTypeSystem(validate = true) {
  const typeSystem = new jsiiReflect.TypeSystem();
  await typeSystem.loadNpmDependencies(path.resolve(__dirname, '..'), {
    validate,
  });
  return typeSystem;
}

export function stackNameFromFileName(fileName: string) {
  return path.parse(fileName).name.replace('.', '-');
}
