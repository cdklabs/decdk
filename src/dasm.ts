import * as cdk from 'aws-cdk-lib';
import chalk from 'chalk';
import yargs from 'yargs';
import { AnnotationsContext, DeclarativeStackError } from './error-handling';
import { CodeEvaluationContext, CodeEvaluator } from './evaluate/code';
import { TypedTemplate } from './type-resolution';
import { loadTypeSystem, readTemplate, stackNameFromFileName } from './util';

let verbosity = 0;

async function main() {
  const argv = await yargs
    .usage(
      '$0 <filename>',
      'Disassemble a CDK app from a declarative JSON or YAML template'
    )
    .option('out', {
      alias: 'o',
      type: 'string',
      demandOption: true,
      describe: 'The directory path to output the disassemble CDK app.',
    })
    .option('verbose', {
      alias: 'v',
      type: 'count',
      description: 'Show debug output. Repeat to increase verbosity.',
    })
    .positional('filename', { type: 'string', required: true }).argv;

  verbosity = argv.verbose;

  const templateFile = argv.filename;
  if (!templateFile) {
    throw new Error('filename is missing');
  }
  const outPath = argv.out;

  const parsedTemplate = await readTemplate(templateFile);
  const stackName = stackNameFromFileName(templateFile);
  const typeSystem = await loadTypeSystem();

  const annotations = AnnotationsContext.root();
  const template = new TypedTemplate(parsedTemplate, {
    annotations,
    typeSystem,
  });

  const context = new CodeEvaluationContext({
    stack: new cdk.Stack(undefined, stackName),
    template,
    typeSystem,
    outPath,
    stackName,
  });
  const ev = new CodeEvaluator(context);

  const project = ev.evaluateTemplate(annotations);

  if (annotations.hasErrors()) {
    throw new DeclarativeStackError(annotations);
  }

  project.saveSync();
}

main().catch((e) => {
  if (e instanceof DeclarativeStackError) {
    e = e.toString(verbosity >= 1);
  }
  // eslint-disable-next-line no-console
  console.error(chalk.red(e));
  process.exit(1);
});
