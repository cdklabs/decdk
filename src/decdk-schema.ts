import yargs from 'yargs';
import { renderFullSchema } from './schema/cdk-schema';
import { loadTypeSystem } from './util';

/* eslint-disable no-console */

async function main() {
  const argv = await yargs
    .usage('$0', 'Generate a JSON Schema to validate deCDK templates')
    .option('warnings', {
      default: true,
      type: 'boolean',
    }).argv;

  const typeSystem = await loadTypeSystem();
  const schema = await renderFullSchema(typeSystem, {
    colors: true,
    warnings: argv.warnings,
    suppressWarnings: [
      // Only an object with methods can satisfy this interface and no built-in class is provided since the feature is intended as an escape hatch
      'aws-cdk-lib.aws_lambda_nodejs.BundlingOptions.commandHooks',
      // Only an object with methods can satisfy this and no built-in class is provided since the feature is intended for user land functionality
      'aws-cdk-lib.BundlingOptions.local',
    ],
  });
  console.log(JSON.stringify(schema, undefined, 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
