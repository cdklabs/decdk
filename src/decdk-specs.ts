import { generateDeCDKSpecs } from './specs/specs-generation';
import { loadTypeSystem } from './util';

/* eslint-disable no-console */

async function main() {
  const typeSystem = await loadTypeSystem();
  const specs = await generateDeCDKSpecs(typeSystem);
  console.log(JSON.stringify(specs, undefined, 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
