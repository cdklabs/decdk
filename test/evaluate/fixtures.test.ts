import path from 'path';
import { readTemplate } from '../../src';
import { testTemplateFixtures, Testing, loadTemplateFixtures } from '../util';

describe('Valid Template Fixtures should synth', () => {
  testTemplateFixtures(async (templateFile) => {
    const template = await readTemplate(templateFile.path);

    const output = await Testing.synth(template);

    expect(template.template).toBeValidTemplate();
    expect(output.template).toMatchSnapshot();
  });
});

describe('Cloudformation templates', () => {
  const cfnFixtures = loadTemplateFixtures([
    path.join(__dirname, '..', 'fixtures/templates/cloudformation'),
  ]);

  const ignoreBecauseCurrentlyFailing: string[] = [
    'cloudformation/condition-same-name-as-resource.json', // There is already a Construct with name 'AlwaysTrue' in DeclarativeStack [Test]
    'cloudformation/condition-using-mapping.json', // does not look like a template
    'cloudformation/find-in-map-for-boolean-property.json', // Expected string or list of strings, got: true
    'cloudformation/find-in-map-with-dynamic-mapping.json', // Expected string, got: {"Ref":"Stage"}
    'cloudformation/fn-sub-escaping.json', //  No resource or parameter with name:  ! DoesNotExist
    'cloudformation/fn-sub-parsing-edges.json', // TypeError: Cannot read properties of null (reading 'Resources')
    'cloudformation/fn-sub-shadow-attribute.json', //  No resource or parameter with name: AnotherBucket
    'cloudformation/functions-and-conditions.json', // Expected list of length 3, got 2
    'cloudformation/hook-code-deploy-blue-green-ecs.json', // Expected valid template, got error(s): -  is not allowed to have the additional property "Hooks"
    'cloudformation/if-in-tags.json', // Expected valid template, got error(s): - Conditions.ValcacheServerEnabled is not of a type(s) object
    'cloudformation/only-parameters-and-rule.json', // does not look like a template
    'cloudformation/outputs-with-references.json', // Expected string, got: {"Ref":"Bucket"}
    'cloudformation/parameter-references.json', // Expected string or list of strings, got: {"Name":"AWS::Include","Parameters":{"Location":{"Ref":"MyParam"}}}
    'cloudformation/resource-attribute-creation-policy.json', // Expected number, got: {"Ref":"CountParameter"}
    'cloudformation/resource-attribute-update-policy.json', // Expected boolean, got: {"Fn::Equals":["true",{"Ref":"WaitOnResourceSignals"}]}
    'cloudformation/short-form-fnsub-string.yaml', // No resource or parameter with name:  ! AWS::Region
  ];

  testTemplateFixtures(
    async (templateFile) => {
      const template = await readTemplate(templateFile.path);
      const output = await Testing.synth(template);
      expect(template.template).toBeValidTemplate();
      expect(output.template).toMatchSnapshot();
    },
    cfnFixtures.filter(({ id }) => !ignoreBecauseCurrentlyFailing.includes(id))
  );
});

describe('Invalid Template Fixtures should fail', () => {
  const invalidFixtures = loadTemplateFixtures([
    path.join(__dirname, '..', 'fixtures/invalid-templates'),
  ]);

  const ignoreBecauseCurrentlyPassing = [
    'invalid-templates/alphabetical-string-passed-to-number.json',
    'invalid-templates/bucket-policy-without-bucket.json',
    'invalid-templates/bucket-with-cors-rules-not-an-array.json',
    'invalid-templates/bucket-with-cors-rules-null-element.json',
    'invalid-templates/bucket-with-invalid-cors-rule.json',
    'invalid-templates/non-existent-condition-in-conditions.json',
    'invalid-templates/non-existent-condition-in-if.json',
    'invalid-templates/non-existent-condition.json',
    'invalid-templates/non-existent-depends-on.json',
    'invalid-templates/non-existent-mapping.json',
    'invalid-templates/non-existent-resource-attribute.json',
    'invalid-templates/only-codecommit-repo-using-cfn-functions.json',
    'invalid-templates/output-referencing-non-existent-condition.json',
    'invalid-templates/rule-referencing-a-non-existent-parameter.json',
    'invalid-templates/short-form-get-att-no-dot.yaml',
    'invalid-templates/short-form-import-sub.yaml',
    'invalid-templates/short-form-transform.yaml',
  ];

  testTemplateFixtures(
    async (templateFile) => {
      const template = await readTemplate(templateFile.path);

      await expect(Testing.synth(template)).rejects.toThrow();
    },
    invalidFixtures.filter(
      ({ id }) => !ignoreBecauseCurrentlyPassing.includes(id)
    )
  );
});

describe('SAM templates', () => {
  const samFixtures = loadTemplateFixtures([
    path.join(__dirname, '..', 'fixtures/templates/sam'),
  ]);

  testTemplateFixtures(async (templateFile) => {
    const template = await readTemplate(templateFile.path);
    const output = await Testing.synth(template);
    expect(template.template).toBeValidTemplate();
    expect(output.template).toMatchSnapshot();
  }, samFixtures);
});

describe('Nested Stacks templates', () => {
  const nestedFixtures = loadTemplateFixtures([
    path.join(__dirname, '..', 'fixtures/templates/nested'),
  ]);

  testTemplateFixtures(async (templateFile) => {
    const template = await readTemplate(templateFile.path);
    const output = await Testing.synth(template);
    expect(template.template).toBeValidTemplate();
    expect(output.template).toMatchSnapshot();
  }, nestedFixtures);
});
