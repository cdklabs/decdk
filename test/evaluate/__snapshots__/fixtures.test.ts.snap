// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested Stacks templates nested/child-import-stack.expected.json 1`] = `
Object {
  "Parameters": Object {
    "MyBucketParameter": Object {
      "Default": "default-bucket-param-name",
      "Type": "String",
    },
    "referencetoAssetParameters5dc7d4a99cfe2979687dc74f2db9fd75f253b5505a1912b5ceecf70c9aefba50S3BucketEAA24F0CRef": Object {
      "Type": "String",
    },
    "referencetoAssetParameters5dc7d4a99cfe2979687dc74f2db9fd75f253b5505a1912b5ceecf70c9aefba50S3VersionKey1194CAB2Ref": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketImport": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": Object {
          "Ref": "MyBucketParameter",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "GrandChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-other-bucket-name",
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "referencetoAssetParameters5dc7d4a99cfe2979687dc74f2db9fd75f253b5505a1912b5ceecf70c9aefba50S3BucketEAA24F0CRef",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "referencetoAssetParameters5dc7d4a99cfe2979687dc74f2db9fd75f253b5505a1912b5ceecf70c9aefba50S3VersionKey1194CAB2Ref",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "referencetoAssetParameters5dc7d4a99cfe2979687dc74f2db9fd75f253b5505a1912b5ceecf70c9aefba50S3VersionKey1194CAB2Ref",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/child-import-stack.json 1`] = `
Object {
  "Parameters": Object {
    "MyBucketParameter": Object {
      "Default": "default-bucket-param-name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketImport": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": Object {
          "Ref": "MyBucketParameter",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "GrandChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-other-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/child-no-bucket.json 1`] = `
Object {
  "Parameters": Object {
    "MyBucketParameter": Object {
      "Default": "default-bucket-param-name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GrandChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-other-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/child-two-parameters.json 1`] = `
Object {
  "Parameters": Object {
    "FirstParameter": Object {
      "Type": "String",
    },
    "SecondParameter": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketImport": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccessControl": Object {
          "Ref": "SecondParameter",
        },
        "BucketName": Object {
          "Ref": "FirstParameter",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "GrandChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "FirstParameter": Object {
            "Ref": "FirstParameter",
          },
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/child-with-number-parameter.yaml 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Parameters": Object {
    "Number": Object {
      "Type": "Number",
    },
  },
  "Resources": Object {
    "S3Bucket": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "testbucket1234unique",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/cross-stack-refs.json 1`] = `
Object {
  "Parameters": Object {
    "Param": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "Bucket": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "Param1": Object {
            "Ref": "Param",
          },
          "Param2": Object {
            "Fn::GetAtt": Array [
              "Bucket",
              "Arn",
            ],
          },
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/child-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/custom-resource.json 1`] = `
Object {
  "Resources": Object {
    "CustomResource": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::CustomResource::Type",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/grandchild-import-stack.json 1`] = `
Object {
  "Parameters": Object {
    "MyBucketParameter": Object {
      "Default": "default-bucket-param-name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketImport": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": Object {
          "Fn::Join": Array [
            "-",
            Array [
              "bucket-name-prefix",
              Object {
                "Ref": "MyBucketParameter",
              },
            ],
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/only-nested-stack.expected.json 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7ArtifactHash605B2835": Object {
      "Description": "Artifact hash for asset \\"6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7\\"",
      "Type": "String",
    },
    "AssetParameters6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7S3BucketC8A1BF52": Object {
      "Description": "S3 bucket for asset \\"6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7\\"",
      "Type": "String",
    },
    "AssetParameters6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7S3VersionKeyA9E03E19": Object {
      "Description": "S3 key for asset version \\"6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "NestedStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParameters6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7S3BucketC8A1BF52",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7S3VersionKeyA9E03E19",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters6b884775090ed88cd1a143f64442a92a6c34eaeff3857976d15ef2e3beee05d7S3VersionKeyA9E03E19",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/only-nested-stack.json 1`] = `
Object {
  "Resources": Object {
    "NestedStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": "doesnt-matter",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-bad-depends-on.json 1`] = `
Object {
  "Resources": Object {
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/child-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-creation-policy.json 1`] = `
Object {
  "Resources": Object {
    "ChildStack": Object {
      "CreationPolicy": Object {},
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-invalid-condition.json 1`] = `
Object {
  "Resources": Object {
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-number-in-child-params.yaml 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Resources": Object {
    "NestedStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "Number": 60,
        },
        "TemplateURL": "https://s3.amazonaws.com/masonme-cdk-test/templates/nested-bucket.yaml",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-one-child.json 1`] = `
Object {
  "Resources": Object {
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-two-children.json 1`] = `
Object {
  "Resources": Object {
    "AnotherChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "another-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/child-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/child-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-two-parameters.json 1`] = `
Object {
  "Resources": Object {
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "FirstParameter": "first-value",
          "SecondParameter": "second-value",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-update-policy.json 1`] = `
Object {
  "Resources": Object {
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdatePolicy": Object {},
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-valid-condition.json 1`] = `
Object {
  "Conditions": Object {
    "AlwaysFalseCond": Object {
      "Fn::Equals": Array [
        Object {
          "Ref": "AWS::Region",
        },
        "completely-made-up-region",
      ],
    },
  },
  "Resources": Object {
    "ChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "MyBucketParameter": "some-magic-bucket-name",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/grandchild-import-stack.json",
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Nested Stacks templates nested/parent-with-attributes.json 1`] = `
Object {
  "Parameters": Object {
    "ArrayParam": Object {
      "Type": "CommaDelimitedList",
    },
  },
  "Resources": Object {
    "AnotherChildStack": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "NotificationARNs": Array [
          "arn1",
        ],
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/child-import-stack.json",
        "TimeoutInMinutes": 5,
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "ChildStack": Object {
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "AnotherChildStack",
      ],
      "Metadata": Object {
        "Property1": "Value1",
      },
      "Properties": Object {
        "NotificationARNs": Object {
          "Ref": "ArrayParam",
        },
        "TemplateURL": "https://cfn-templates-set.s3.amazonaws.com/child-import-stack.json",
        "TimeoutInMinutes": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Ref": "ArrayParam",
            },
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
